src/main/scala/beam/utils/data/ctpp/JointDistribution.scala

```diff
@@ -0,0 +1,148 @@
+package beam.utils.data.ctpp
+
+import java.util.{Map => JavaMap}
+
+import org.apache.commons.math3.util.{Pair => CPair}
+import beam.utils.csv.GenericCsvReader
+import beam.utils.data.ctpp.JointDistribution.{CustomRange, RETURN_COLUMN}
+import org.apache.commons.math3.distribution.EnumeratedDistribution
+import org.apache.commons.math3.random.RandomGenerator
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+object JointDistribution extends GenericCsvReader {
+
+  case class CustomRange(start: Double, end: Double)
+
+  val RANGE_COLUMN_TYPE = "range"
+  val DOUBLE_COLUMN_TYPE = "double"
+  val INT_COLUMN_TYPE = "int"
+  val STRING_COLUMN_TYPE = "string"
+  val RETURN_COLUMN = "probability"
+
+  def fromCsvFile(
+    pathToCsv: String,
+    rndGen: RandomGenerator,
+    columnMapping: Map[String, String] = Map(),
+    scale: Boolean = false
+  ): JointDistribution = {
+    def toScalaMap(rec: JavaMap[String, String]): Map[String, String] = rec.asScala.toMap
+    val (it, toClose) = readAs[Map[String, String]](pathToCsv, toScalaMap, _ => true)
+    val mappedArray: Array[Map[String, String]] = try {
+      it.toArray
+    } finally {
+      Try(toClose.close())
+    }
+
+    val detectedColumn = mappedArray(0).map {
+      case (key, value) => (key, if (value.contains(",")) RANGE_COLUMN_TYPE else STRING_COLUMN_TYPE)
+    }
+    if (columnMapping.nonEmpty)
+      new JointDistribution(mappedArray, rndGen, columnMapping, scale)
+    else
+      new JointDistribution(mappedArray, rndGen, detectedColumn, scale)
+  }
+}
+
+/**
+  *
+  * @param mappedArray
+  * @param rndGen Random number generator. If an instance of `JointDistribution` is shared across multiple threads,
+  *               make sure you use thread-safe instance of `RandomGenerator`, for example, `SynchronizedRandomGenerator`
+  *               or wrap it by your own implementation which uses ThreadLocal
+  * @param columnMapping
+  * @param scale
+  */
+class JointDistribution(
```

Comment:
@rajnikantsh please, add some comments to the constructor explaining what are those input parameters

Reply:
@REASY now we have added details on constructor inputs